yaxis tick formatters

ppisljar opened the pull request:
Ability to specify one of (bits,bits/s,bytes,bytes/s,currency,custom) unit formatters for y-axis labels.

based on elastic/timelion#99 thanks and credits to @pagenbag

Example below of :
```
.label("series1:bandwidth").yaxis(units=bits/s,max=2000000)
.label("series2:currency").yaxis(2,units=currency:EUR,max=2000000)
.label("series3:custom").yaxis(3,position=right,units="custom:~ : things",max=2000000)
```

![6f199076-0254-11e6-816c-70e7b1e03e63 1](https://cloud.githubusercontent.com/assets/13629809/20268708/981f2a9c-aa80-11e6-8481-c9714816eb0d.png)

rashidkpc commented:
I can't get the `custom` format to work at all, can you? I just get a blank chart

![screen shot 2016-11-14 at 10 21 11 am](https://cloud.githubusercontent.com/assets/1250387/20275021/19722702-aa54-11e6-8285-b6343b480814.png)

The currency formatting functionality is confusing. I can't get the example in the screenshot to work and the spec for toLocaleString says you need to know the 3 letter ISO 4217 currency code. Perhaps specifying the symbol, as in the screenshot, is a non-standard thing implemented in some other browser than Chrome?

ppisljar commented:
oh sorry, the screenshot is misleading. you need to provide 3 letter ISO code that's correct. i changed that from original implementation.

i fixed the custom format.

rashidkpc commented:
Should probably note in the help text for the `units` parameter that the currency requires the 3 letter ISO 4217 currency code. Or maybe just get rid of the currency options and just use `'custom:$'`. I suspect that would be sufficient for most people?

ppisljar commented:
i like that currency also adds the commas on the thousands and limits decimal places to 2 ... i'll add the information to help text.

rashidkpc commented on src/core_plugins/timelion/public/services/tick_formatters.js:
```diff
@@ -0,0 +1,53 @@
+      'bits/s': function (val, axis) {
+        var labels = ['b/s','kb/s','mb/s','gb/s','tb/s','pb/s'];
+        var index = 0;
+        while (val > 1000 && index < labels.length) {
```
`val > 1000` should probably be `val >= 1000` otherwise you end up with stuff like this:

![screen shot 2016-11-14 at 12 45 52 pm](https://cloud.githubusercontent.com/assets/1250387/20280925/b4cfc2ba-aa6b-11e6-8323-275fc5c6a6bd.png)

rashidkpc commented:
The bit above lead me to another issue. When using the `bytes` mode the ticks end up placed at unnatural points:

Currently:
![screen shot 2016-11-14 at 1 12 31 pm](https://cloud.githubusercontent.com/assets/1250387/20281078/4970a402-aa6c-11e6-9db3-64d2e6389236.png)

What I would expect:
![screen shot 2016-11-14 at 1 13 35 pm](https://cloud.githubusercontent.com/assets/1250387/20281117/65c78940-aa6c-11e6-8e13-46476abddbcc.png)

The logic for the 2nd chart can be cribbed from https://github.com/whatbox/jquery.flot.byte/blob/master/jquery.flot.byte.js

I wouldn't necessarily use that plugin, but it does the right thing.

ppisljar commented:
i took the logic for generating byte ticks from the plugin you pointed me to. i kept just the basic part we need.

pagenbag commented:
Yay - thanx @ppisljar :)

rashidkpc commented:
You need to take `['bits, bits/s']` out of the modes to pass to the tick formatter, otherwise you end up with stuff like this on the chart on the left.

![screen shot 2016-11-16 at 10 02 46 am](https://cloud.githubusercontent.com/assets/1250387/20357198/3e528f9e-abe4-11e6-8f99-14feee5df33e.png)

What you want is this:
![screen shot 2016-11-16 at 9 56 09 am](https://cloud.githubusercontent.com/assets/1250387/20357215/4c8dd320-abe4-11e6-85c0-6b0857c58529.png)

Also, as @simianhacker noted to me, you can get rid of all of the code in `tick_formatters.js` and replace it with calls to numeral.js which is already ships with Kibana. Here's how @simianhacker is doing tickFormatters: https://github.com/elastic/thor/blob/master/public/lib/tick_formatter.js

rashidkpc commented on src/core_plugins/timelion/server/series_functions/yaxis.js:
```diff
@@ -58,7 +74,13 @@ module.exports = new Chainable('yaxis', {
+        if (!tickFormatters[unitTokens[0]]) {
+          throw new Error (`${units} is not a supported unit type.`);
+        }
+        myAxis.units = unitTokens;
```
For the sake of API consistency and readability of the returned object, I'd rather see this returned as 

```
{
  "type": "someType",
  "prefix": "foo",
  "suffix": "bar"
}
```

ppisljar commented:
i fixed most things you brought up but:
- there is a [bug with secondary y axes](https://github.com/elastic/kibana/issues/9114) in timelion which i think we should address in separate PR
- there's [some](https://github.com/elastic/kibana/issues/8583) [discussion](https://github.com/elastic/kibana/issues/8584) going on about depending on numeraljs for user entered information ... for now i would prefer not to depend on it.

ppisljar commented:
timelion doesnt have any tests for the browser part does it ?

rashidkpc commented:
- we should fix that one first, it sounds like the 2nd y-axis is partially broken, we don't want it completely broken
- Both of those tickets are about numeral.js as a user-facing API. This wouldn't directly expose numeral to the user, it would be an implementation detail, as numeral should be. 

Timelion doesn't have tests for the browser stuff, but pulling out that function and adding tests for it shouldn't be much work. If you take it out of the file you can test it in isolation. There's a fair amount of logic there.

ppisljar commented:
hey @rashidkpc  yes tests are there, also i fixed the axis (its still partly broken due to that bug, but its not completely broken).

with numeral.js i don't seem to be able to do bits, bits/s, bytes/s and currency (in a way i do it now, where you just submit ISO currency code and it will auto format to match ... with numeral you can just add a sign there (which is tricky for chinese for example ... where on my keyboard is CNY sign ? ) ) or am i just not getting something ?

rashidkpc commented:
This now seems to be broken completely unless I call at least one series with `.yaxis(#)`. Rendering fails relatively silently, only logging a console message, but the chart isn't redrawn without `.yaxis(#)`

Also the currency formatter is currently broken. The `tick_formatters.js` file needs tests. You're referencing `axis.options.prefix` in the formatter, but its actually under `axis.options.units.prefix` 

On a side note, its funny you mention the Chinese yuan, since its an excellent example of lack of prefix standardization in currency. 

In China you would see ¥10.00, however the rest of the world uses '¥' to refer to the Japanese yen. The official name of the yuan is of course the renminbi, and you'll often see it abbreviated as RMB outside of China. With this code we get Javascript's `toLocaleString()` output: CN¥10.00, which is a weird mashup, especially for web browsers. By default they treat ¥ as word break boundary and will wrap CN and ¥10.00 onto different lines. 

In any case, ISO 4217 provides a standard for something, so it'll have to do.

rashidkpc commented:
Everything about this looks fine, with one exception. I'd like it to throw an error if the currency isn't either null (thus defaults to USD) or a 3 letter code, eg `/[A-Za-z]{3}/`, otherwise the user just gets a blank chart.

rashidkpc commented:
Well this is weird...

Seems to only happen with the custom formatter and only when the data is all between 0 and 1. Passing .precision(1) doesn't seem to fix it either. I swear I'm not using an intel processor from the early 90s, so a buggy FPU is out of the question 😄 

![screen shot 2016-12-02 at 9 20 04 am](https://cloud.githubusercontent.com/assets/1250387/20841432/018a0044-b871-11e6-8e67-86e4738fd0cd.png)

ppisljar commented:
@thomasneirynck i cant reproduce it either. i followed the same pattern rest of timelion uses to throw errors and they usually don't show up in server log but in the client.

@rashidkpc can you reproduce with static line ? something like 
`.static(0.000000000000000000454352422).yaxis(units=custom:FOO:BAR)` works ok for me

ppisljar commented:
jenkins, test this

rashidkpc commented:
@ppisljar You happened to pick one of the rare values that doesn't cause the issue:

![weird](https://cloud.githubusercontent.com/assets/1250387/21364953/52c18f7a-c6b0-11e6-834c-6b9ffba38ed5.gif)

I'm not sure if this is even fixable, but shipping with this bug seems really bad since it happens with 0, which is pretty common

<img width="1005" alt="screen shot 2016-12-20 at 12 33 13 pm" src="https://cloud.githubusercontent.com/assets/1250387/21364999/83e149ce-c6b0-11e6-9f7f-9b818f15cb7f.png">

ppisljar commented:
i was able to reproduce the bug ... but funny, not with same numbers as you ... for example with 0 it works.

what about removing the custom option ? (its the only one causing the issue)

ppisljar commented:
its fixed now @rashidkpc  please take another look.
i finished doing the same flot is doing to make sure the numbers look fine.
